sdk/src/transaction_context.rs

```diff
@@ -215,11 +227,21 @@ impl TransactionContext {
 
     /// Pops the current InstructionContext
     pub fn pop(&mut self) -> Result<(), InstructionError> {
-        if self.instruction_stack.is_empty() {
-            return Err(InstructionError::CallDepth);
-        }
+        let instruction_context = self.get_current_instruction_context()?;
+        let detected_unbalanced_instruction = if self.rent.is_some() {
+            let instruction_accounts_lamport_sum =
+                self.instruction_accounts_lamport_sum(&instruction_context.instruction_accounts)?;
+            // Verify that the total sum of all lamports in this instruction did not change
+            instruction_context.instruction_accounts_lamport_sum != instruction_accounts_lamport_sum
+        } else {
+            false
+        };
```

Comment:
maybe, this comment might help future readers:

```rust
// always `.pop()` even if we detected unblanced instruction to avoid unmatched .push() and .pop()
self.instruction_stack.pop();
```